classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+        $mergerequest = $DB->get_record(merge_request::TABLE_MERGE_REQUEST,
+                                            ['id' => $data->mergerequestid]);
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
+        $this->verify_users_to_keep_and_remove($mergerequest);
+        $mergerequestresult = $this->merge($mergerequest, $maxattempts, merge_request::TRIED_WITH_ERROR);
+        if ($mergerequestresult->status == merge_request::COMPLETED_WITH_SUCCESS) {
+            /* We run the merge request AGAIN because the user may be interacting with Moodle
+            * while merge request is being processed, so that NO ALL records are correctly migrated
+            * into the user to keep. It will ensure ALL records are correctly migrated into the user to keep. */
+            $mergerequestresult = $this->merge($mergerequestresult, $maxattempts, merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequestresult->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequestresult->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(object $record, int $maxattempts, int $statusforerror) {
+        $retries = $record->retries + 1;
+        // Update retries.
+        $this->update_retries_in_table($record->id, $retries);
+        $logs = [];
+        $logs = $record->log;
+        $mut = new MergeUserTool();
+        list($success, $log) = $mut->merge($record->keepuserid, $record->removeuserid);
+        if (!$success) {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        } else {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        }
+        // Append logs to the list.
+        // $baseitem->log[$item->timemodified] = json_decode($item->log, false);
+        $logs[$record->timeadded] = $log;
+        $this->update_status_and_log_in_table($record->id,
+                                            $status,
+                                            $logs);
+        $record->log = $logs;
+        $record->status = $status;
+        return $record;
+    }
+    /**
+     * Function for updating the status of the merging request.
+     */
+    private function update_status_and_log_in_table(int $idrecord,
+                                                    int $status,
+                                                    array $log): void {
+        global $DB;
+        if ($status == merge_request::COMPLETED_WITH_SUCCESS ||
```

Comment:
I would do something like this:
```php
$update = (object)[
                'id' => $idrecord,
                'status' => $status,
                'log' => json_encode($log),
            ];
if ($status == merge_request::COMPLETED_WITH_SUCCESS ||
$status == merge_request::COMPLETED_WITH_ERRORS) {
    $update->timecompleted = time();
}
```

This way we prevent repeated code and make it simple.

Reply:
format it properly, please, hehehe

Reply:
done

Reply:
done.